src/com/seafile/seadroid2/cameraupload/CameraSyncAdapter.java

```diff
@@ -0,0 +1,686 @@
+                continue;
+            }
+
+            if (dbHelper.isUploaded(file)) {
+                Log.d(DEBUG_TAG, "Skipping media " + file + " because we have uploaded it in the past.");
+                continue;
+            }
+
+            uploadFile(dataManager, file, bucketName);
+        }
+
+        waitForUploads();
+        checkUploadResult(syncResult);
+    }
+
+    private void waitForUploads() throws InterruptedException {
+        Log.d(DEBUG_TAG, "wait for transfer service to finish our tasks");
+        WAITLOOP: while (!isCancelled()) {
+            Thread.sleep(100); // wait
+
+            for (int id: tasksInProgress) {
+                UploadTaskInfo info = txService.getUploadTaskInfo(id);
+                if (info.state == TaskState.INIT || info.state == TaskState.TRANSFERRING) {
+                    // there is still at least one task pending
+                    continue  WAITLOOP;
+                }
+            }
+            break;
+        }
+    }
+
+    /**
+     * Upload is finished. Go through task infos and mark files as uploaded in our DB
+     *
+     * @param syncResult
+     * @throws SeafException
+     */
+    private void checkUploadResult(SyncResult syncResult) throws SeafException {
+        for (int id: tasksInProgress) {
+            UploadTaskInfo info = txService.getUploadTaskInfo(id);
+            if (info.err != null) {
+                throw info.err;
+            }
+            if (info.state == TaskState.FINISHED) {
+                File file = new File(info.localFilePath);
+                dbHelper.markAsUploaded(file);
+                syncResult.stats.numInserts++;
+            } else {
+                throw SeafException.unknownException;
+            }
+        }
+    }
+
+    /**
+     * Upload a media file to the seafile server.
+     *
+     * @param dataManager Handle to the seafile server
+     * @param file        the file to be uploaded
+     * @param bucketName  the name of the media bucket
+     * @throws SeafException
+     */
+    private void uploadFile(DataManager dataManager, File file, String bucketName) throws SeafException {
+
+        String serverPath = Utils.pathJoin(targetDir, BASE_DIR, bucketName);
+
+        List<SeafDirent> list = dataManager.getCachedDirents(targetRepoId, serverPath);
+        if (list == null) {
+            Log.e(DEBUG_TAG, "Seadroid dirent cache is empty in uploadFile. Should not happen, aborting.");
+            // the dirents were supposed to be refreshed in createDirectories()
+            // something changed, abort.
+            throw SeafException.unknownException;
+        }
+
+        /*
+         * We don't want to upload a file twice unless the local and remote files differ.
+         *
+         * It would be cool if the API2 offered a way to query the hash of a remote file.
+         * Currently, comparing the file size is the best we can do.
```

Comment:
I think media files are seldom changed comparing with other types of files. Comparing the hash value of both local and remote is a good way to detect changes. Unfortunately, we such hash value does not exist. Though calling the API http://manual.seafile.com/develop/web_api.html#upload-file-1will return a hash value, it is invalid to use it to track local remote changes, what\`s more, it is cpu density operation which may slow performance.